hibernate-core/src/main/java/org/hibernate/query/internal/QueryParameterBindingsImpl.java

```diff
@@ -167,30 +173,49 @@ public void visitBindings(@SuppressWarnings("rawtypes") BiConsumer action) {
 	}
 
 	@Override
-	public QueryKey.ParameterBindingsMemento generateQueryKeyMemento(SharedSessionContractImplementor persistenceContext) {
-		final MutableCacheKeyImpl mutableCacheKey = new MutableCacheKeyImpl(parameterBindingMap.size());
-
+	public QueryKey.ParameterBindingsMemento generateQueryKeyMemento(SharedSessionContractImplementor session) {
+		// Create separate cache keys for each parameter binding
+		final List<MutableCacheKeyImpl> cacheKeys = new ArrayList<>();
 		for ( Map.Entry<QueryParameter<?>, QueryParameterBinding<?>> entry : parameterBindingMap.entrySet() ) {
 			final QueryParameterBinding<?> binding = entry.getValue();
 			final MappingModelExpressible<?> mappingType = determineMappingType(
 					binding,
 					entry.getKey(),
-					persistenceContext
+					session
 			);
-
+			final MutableCacheKeyImpl cacheKey;
 			if ( binding.isMultiValued() ) {
+				cacheKey = new MutableCacheKeyImpl( binding.getBindValues().size() );
 				for ( Object bindValue : binding.getBindValues() ) {
 					assert bindValue != null;
-					mappingType.addToCacheKey( mutableCacheKey, bindValue, persistenceContext );
+					mappingType.addToCacheKey( cacheKey, bindValue, session );
 				}
 			}
 			else {
+				cacheKey = new MutableCacheKeyImpl( 1 );
 				final Object bindValue = binding.getBindValue();
-				mappingType.addToCacheKey( mutableCacheKey, bindValue, persistenceContext );
+				mappingType.addToCacheKey( cacheKey, bindValue, session );
 			}
+			cacheKeys.add( cacheKey );
 		}
-
-		return mutableCacheKey.build();
+		// Add any enabled filter parameter values to the cache key
+		for ( String filterName : session.getLoadQueryInfluencers().getEnabledFilterNames() ) {
+			final FilterImpl filter = (FilterImpl) session.getLoadQueryInfluencers().getEnabledFilter( filterName );
+			final FilterDefinition filterDefinition = filter.getFilterDefinition();
+			for ( String paramName : filterDefinition.getParameterNames() ) {
+				final Object paramValue = filter.getParameter( paramName );
+				final JdbcMapping jdbcMapping = filterDefinition.getParameterJdbcMapping( paramName );
+				final MutableCacheKeyImpl cacheKey = new MutableCacheKeyImpl( 1 );
+				addBasicValueToCacheKey( cacheKey, paramValue, jdbcMapping, session );
+				cacheKeys.add( cacheKey );
+			}
+		}
+		// Sort the parameter cache key list using each key's hash code to ensure consistency
+		cacheKeys.sort( Comparator.comparingInt( k -> k.hashCode ) );
```

Comment:
This shouldn't be necessary. We should only need a single `MutableCacheKeyImpl` like before. I think the problems you might have encountered are because we don't add anything to `MutableCacheKeyImpl` for `null` values, which is a problem in general.

Reply:
I think it is necessary, how can we be sure that the parameter values are always added to the cache key in the same order? The reproducer for HHH-16594 causes parameters to be in different orders each time the criteria query is executed.

Reply:
I think we have to make sure that `parameterBindingMap` is a `LinkedHashSet`, then we have a fully deterministic ordering which we would need for replicated/clustered caches anyway.

Reply:
Updated the solution, now we ensure the parameters are added to the cache key in the order they were processed through a custom `LinkedIdentityHashMap` and there's no need of manual sorting.